File: junior/contest4/prog/graph.java
import java.io.*;
import java.util.*;

public class graph {
	private static void mul(int[][] A, int[][] B, int[][] C) {
		for(int m = 0; m < 10; ++m) {
			for(int p = 0; p < 10; ++p) {
				int sum = 0;
				for(int n = 0; n < 10; ++n)
					sum += A[m][n] * B[n][p];
				C[m][p] = sum;
			}
		}
	}

	public static void main(String[] args) throws IOException {
		Scanner input = new Scanner(System.in);
		for(int n = 0; n < 5; ++n) { // read 5 input cases
			int[][] ma = new int[10][10]; // default to all zeros
			String line = input.nextLine();
			int prop = line.charAt(0) - '0';
			for(int i = 2; i < line.length();) {
				int a = line.charAt(i++) - '0';
				int b = line.charAt(i++) - '0';
				ma[a][b] = 1;
				i++; // skip a space 
			}
			switch(prop) {
			case 1:
				{
					int res = 0;
					for(int i = 1; i < 10; ++i) {
						res += ma[i][i]; // length 1 cycles
						for(int j = i+1; j < 10; ++j)
							res += ((ma[i][j] == 1 && ma[j][i] == 1) ? 1 : 0);
					}
					System.out.println(res);
				}
				break;
			case 2:
				{
					int max_edges = 0;
					int max_sum = 0;
					for(int i = 1; i < 10; ++i) {
						int edges = 0;
						int sum = 0;
						for(int j = 1; j < 10; ++j) {
							edges += ma[i][j];
							if(ma[i][j] == 1)
								sum += (i*10+j);
						}
						if(edges > max_edges) {
							max_edges = edges;
							max_sum = sum;
						}
					}
					System.out.println(max_sum);
				}
				break;
			case 3:
				{
					int[][] mc = new int[10][10];
					mul(ma, ma, mc);
					int res = 0;
					for(int i = 0; i < 10; ++i)
						for(int j = 0; j < 10; ++j)
							res += mc[i][j];
					System.out.println(res);
				}
				break;
			}
		}
	}
}

File: 2020-21/contest4/prog/junior/graph.py
#!/usr/bin/python3
import sys
import math

# AB = C
def multiply(N,A,B,C):
	for m in range(N):
		for p in range(N):
			total = 0
			for n in range(N):
				total += A[m][n] * B[n][p]
			C[m][p] = total

if __name__ == "__main__":
	for n in range(0,5):
		# Create a 10x10 matrix with all zeros in one line
		ma = [[0]*10 for i in range(10)]
		# Read all numbers into a list of strings
		nums = [x for x in input().split()]
		prop = int(nums[0])
		# Create the adjacency matrix
		for i in range(1,len(nums)):
			ma[int(nums[i][0])][int(nums[i][1])] = 1
		if prop == 1:
			res = 0
			for i in range(1,10):
				res += ma[i][i] # length of 1 cycles
				for j in range(i+1,10): # length of 2 cycles
					res += (1 if ma[i][j] == 1 and ma[j][i] == 1 else 0)
			print(res)
		elif prop == 2:
			max_edges = 0
			max_total = 0
			for i in range(1,10):
				edges = 0
				total = 0
				for j in range(1,10):
					edges += ma[i][j]
					if ma[i][j] == 1:
						total += (i*10+j)
				if edges > max_edges:
					max_edges = edges
					max_total = total
			print(max_total)
		elif prop == 3:
			mc = [[0]*10 for i in range(10)] # DO NOT DO THIS -> [[0]*10]*10
			multiply(10,ma,ma,mc)
			res = 0
			for i in range(1,10):
				for j in range(1,10):
					res += mc[i][j]
			print(res)
